src/ProxyKit.Tests/EndToEndTests.cs

```diff
@@ -1,62 +1,231 @@
+                    result.StatusCode.ShouldBe(HttpStatusCode.BadRequest);
+                    result = await client.GetAsync("/realServer/error");
+                    result.StatusCode.ShouldBe(HttpStatusCode.InternalServerError);
+
+                    // server timeouts should be returned as gateway timeouts
+                    result = await client.GetAsync("/realServer/slow");
+                    result.StatusCode.ShouldBe(HttpStatusCode.GatewayTimeout);
+
+                    // server timeouts should be 'delayed' 
+                    using (var cts = new CancellationTokenSource())
+                    {
+                        cts.CancelAfter(TimeSpan.FromMilliseconds(1000));
+                        try
+                        {
+                            result = await client.GetAsync("/realServer/slow", cts.Token);
+                            result.StatusCode.ShouldBe(HttpStatusCode.GatewayTimeout);
+                        }
+                        catch (OperationCanceledException)
+                        {
+                        }
+                    }
+                    // When server is stopped, should return 
+                    await server.StopAsync();
+                    result = await client.GetAsync("/realServer/normal");
+                    result?.StatusCode.ShouldBe(HttpStatusCode.ServiceUnavailable);
+                }
+            }
+
+        }
+
+
+        [Fact]
+        public async Task When_timeout_is_really_tight_and_server_not_available_then_operation_cancelled_is_handled_correctly()
         {
-            var client = _testServer.CreateClient();
-            var result = await client.GetAsync("/accepted");
-            result.StatusCode.ShouldBe(HttpStatusCode.Accepted);
+            using (var server = BuildKestrelBasedServerOnRandomPort())
+            {
+                await server.StartAsync();
+                var port = GetServerPort(server);
+
+                using (var testServer = new TestServer(new WebHostBuilder()
+                    .UseSetting("port", port.ToString())
+                    .UseSetting("timeout", "1")
+                    .UseStartup<TestStartup>()))
+                {
+                    var client = testServer.CreateClient();
+                    // When server is running, response code should be 'ok'
+                    var result = await client.GetAsync("/realServer/normal");
+                    result.StatusCode.ShouldBe(HttpStatusCode.OK);
+
+                    // When server is stopped, should return 
+                    await server.StopAsync();
+                    result = await client.GetAsync("/realServer/normal");
+                    if (result != null)
+                    {
+                        try
+                        {
+                            result.StatusCode.ShouldBe(HttpStatusCode.GatewayTimeout);
+                        }
+                        catch (NullReferenceException)
```

Comment:
@osoykan Strangely enough, this test fails with a null reference exception on the build server.. But I don't see why.. do you have any thoughts?

Reply:
As a result of running the build pipeline in a linux docker container; it shows us the solution will be:
https://github.com/osoykan/ProxyKit/blob/patch-1/src/ProxyKit/HttpContextExtensions.cs#L56-L73

Sent a PR: https://github.com/Erwinvandervalk/ProxyKit/pull/2